src/rust/lib/graph-editor/src/component/operator.rs

```diff
@@ -0,0 +1,271 @@
+        }
+    }
+}
+
+
+
+// ===========================
+// === Fullscreen Operator ===
+// ===========================
+
+pub trait Fullscreenable = display::Object+Resizable+NativeComponent+Networked+CloneRef+'static;
+
+/// A `FullscreenOperator` can be used to apple fullscreen mode to a UI element as well as undo the
+/// undo the fullscreen operation and restore the previous state. The  `FullscreenOperator` can be
+/// applied to any target that implements `display::Object`, `Resizable` and `NativeUiElement`.
+// TODO consider incorporating these traits into display::Object or another common "SceneElement"
+// type. But it is important that complex UI components can provide information about their
+// sub-components (for example, multiple sub-shapes or HTML components).
+#[derive(Debug)]
+pub struct FullscreenOperator<T> {
+    target            : T,
+    scene             : Scene,
+    size_original     : Vector3<f32>,
+    position_original : Vector3<f32>,
+    parent_original   : Option<display::object::Instance>,
+}
+
+impl<T:Fullscreenable> FullscreenOperator<T> {
+    /// Make the provided target fullscreen within the given scene and return the
+    /// `FullscreenOperator`.
+    pub fn apply(target:T, scene:Scene) -> Self {
+        let size_original     = target.size();
+        let position_original = target.position();
+        let parent_original   = target.display_object().rc.parent();
+        FullscreenOperator {target,scene,size_original,position_original,parent_original}.init()
+    }
+
+    fn init(self) -> Self {
+        let original_pos = self.target.display_object().global_position();
+
+        // Change parent
+        self.target.display_object().set_parent(self.scene.display_object());
+        self.target.unset_layers_all(&self.scene);
+        set_layers_fullscreen(&self.target, &self.scene);
+
+        let margin = 0.0;
+        let scene_shape = self.scene.shape();
+        let size_new    = Vector3::new(scene_shape.width(), scene_shape.height(),0.0) * (1.0 - margin);
+
+        // TODO Currently we assume objects are center aligned, but this needs to be properly
+        // accounted for here.
+
+        let frp_network      = &self.target.network().clone_ref();
+        let target_pos       = Vector3::zero();
+        let original_size    = self.size_original;
+        let target_size      = size_new;
+        let target           = self.target.clone_ref();
+        let resize_animation = animation(frp_network, move |value| {
+            let pos  = original_pos  * (1.0 - value) + target_pos * value;
+            let size = original_size * (1.0 - value) + target_size * value;
+            target.set_position(pos);
+            target.set_size(size);
+        });
+        resize_animation.set_target_position(1.0);
+
+        self.scene.views.toggle_overlay_cursor();
+
+        self
+    }
+
+    /// Undo the fullscreen operation and restore the previous state exactly as it was.
+    pub fn undo(self) {
```

Comment:
What does `my_object.undo()` does? If this is designed to be a method, it should be named `undo_fullscreen`.